pkg/controllers/provisioning/v1alpha1/reallocation/annotate.go

```diff
@@ -0,0 +1,97 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reallocation
+
+import (
+	"context"
+	"fmt"
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha1"
+	utilsnode "github.com/awslabs/karpenter/pkg/utils/node"
+	"go.uber.org/zap"
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"time"
+)
+
+type Annotator struct {
+	kubeClient client.Client
+}
+
+// Takes in a list of underutilized nodes and populates the nodes with the ProvisionerUnderutilized Label
+func (a *Annotator) LabelUnderutilized(ctx context.Context, nodes []*v1.Node) {
+	for _, node := range nodes {
+		persisted := node.DeepCopy()
+		node.Labels[v1alpha1.ProvisionerUnderutilizedKey] = "true"
+		if err := a.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+			zap.S().Debugf("Continuing after failing to patch node %s, %s", node.Name, err.Error())
+		}
+	}
+}
+
+// AddTTLs takes in a list of underutilized nodes and adds TTL to them
+func (a *Annotator) AddTTLs(ctx context.Context, nodes []*v1.Node) error {
+	for _, node := range nodes {
+		persisted := node.DeepCopy()
+		node.Annotations[v1alpha1.ProvisionerTTLKey] = time.Now().Add(300 * time.Second).Format(time.RFC3339)
+		if err := a.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+			return fmt.Errorf("patching node, %w", err)
+		}
+		zap.S().Debugf("Added TTL to node %s", node.Name)
+	}
+	return nil
+}
+
+// RemoveUtilizedTTLs takes in a list of nodes labeled underutilized and removes TTL if there is sufficient resource usage
+func (a *Annotator) RemoveUtilizedTTLs(ctx context.Context, nodes []*v1.Node) {
+	for _, node := range nodes {
+		pods, err := a.getPodsOnNode(ctx, node.Name)
+		if err != nil {
+			zap.S().Debugf("Ignoring error: could not get pods on node %s, %s", node.Name, err.Error())
+			continue
+		}
+		if !utilsnode.IsUnderutilized(pods) {
+			persisted := node.DeepCopy()
+			delete(node.Labels, v1alpha1.ProvisionerUnderutilizedKey)
+			delete(node.Annotations, v1alpha1.ProvisionerTTLKey)
+			if err := a.kubeClient.Patch(ctx, node, client.MergeFrom(persisted)); err != nil {
+				zap.S().Debugf("Continuing after failing to remove TTL from node %s, %s", node.Name, err.Error())
```

Comment:
I think if we can't patch the node, we should just return an error since something is wrong. The reconcile loop will try again and hopefully won't run into issues again (retriable errors)

Reply:
Ok that sounds good to me. I was a little concerned I wasn't handling errors here, but wanted to make sure that we are not limiting the rest of the nodes if we fail.